### PR TITLE
Add notebook for iceberg spark

### PR DESCRIPTION
**Pull request by rymurr**
This PR adds a new notebook for Iceberg and Spark using Iceberg `0.12.0` and the new spark binder notebook. It also adds the Nessie SQL extensions and removes most of the usage of the command line interface.

There is some scaffolding which is a copy of the scaffolding in `notebook-tests` as this will eventually replace that directory.

---
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie-demos/144)

### CODE REVIEW

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 1]
```diff
@@ -0,0 +1,619 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Nessie Spark SQL Demo with NBA Dataset\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:7]: No clue whether and how easy it is, but would it be possible to have a notebook that uses the `spark-sql` REPL? Not asking for that for this PR though.

File: .github/workflows/notebooks.yaml [hunk 2]
```diff
@@ -66,3 +66,6 @@ jobs:
       - name: Test with tox
         working-directory: notebook-tests
         run: tox
+      - name: Test Notebooks with Tox
+        working-directory: notebooks/tests
```
Comment (snazy) [inline .github/workflows/notebooks.yaml:70]: I suspect, notebook-tests will go away, right?

File: notebooks/tox.ini [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# limitations under the License.
+#
+
+[tox]
+envlist = py37, py38, py39, flake8
+skipsdist = True
+
+[gh-actions]
+python =
+    3.7: py37
```
[the remark this answers is not in the document]
Comment (rymurr) [reply notebooks/tox.ini:23]: agreed, fixed

File: notebooks/tox.ini [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+[testenv]
+setenv =
+    PYTHONPATH = {toxinidir}
+    NESSIE_DEMO_ROOT = file://{toxinidir}/../
+passenv = TOXENV CI TRAVIS TRAVIS_* CODECOV_* NESSIE_*
+deps =
+    -r{toxinidir}/requirements_dev.txt
+    -r{toxinidir}/requirements.txt
+    -r{toxinidir}/../binder/requirements.txt
```
[the remark this answers is not in the document]
Comment (rymurr) [reply notebooks/tox.ini:46]: it doesn't matter, just in case have moved it

File: notebooks/tox.ini [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    -r{toxinidir}/requirements.txt
+commands =
+    flake8 tests
+    safety check
+    mypy --ignore-missing-imports -p tests
+
+[testenv]
+setenv =
+    PYTHONPATH = {toxinidir}
+    NESSIE_DEMO_ROOT = file://{toxinidir}/../
```
[the remark this answers is not in the document]
Comment (rymurr) [reply notebooks/tox.ini:41]: removed

File: notebooks/tox.ini [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+passenv = TOXENV CI TRAVIS TRAVIS_* CODECOV_* NESSIE_*
```
[the remark this answers is not in the document]
Comment (rymurr) [reply notebooks/tox.ini:42]: removed

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import findspark\n",
+    "from pyspark.sql import *\n",
+    "from pyspark import SparkConf\n",
+    "findspark.init()\n",
+    "\n",
+    "conf = SparkConf()\n",
+    "conf.set(\"spark.jars.packages\", \"org.apache.iceberg:iceberg-spark3-runtime:0.12.0,org.projectnessie:nessie-spark-extensions:0.9.0\") # we need iceberg libraries and the nessie sql extensions\n",
```
[the remark this answers is not in the document]
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:38]: done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 1 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:7]: its possible, but a bit more of a faf to set up. Definitely for a follow up

File: .github/workflows/notebooks.yaml (hunk 2 above)
Comment (rymurr) [reply .github/workflows/notebooks.yaml:70]: correct

File: notebooks/tests/__init__.py [hunk 8]
```diff
@@ -61,24 +62,29 @@ def _get_unzip(filename: str, url: str) -> None:
...
     _get_hadoop()
     if not os.getenv("HADOOP_HOME"):
-        raise Exception("The HADOOP_HOME env var must be set and point to a valid Hadoop installation")
+        raise Exception(
+            "The HADOOP_HOME env var must be set and point to a valid Hadoop installation"
+        )
 
     pyflink_lib_dir = _find_pyflink_lib_dir()
-    for i, jar in enumerate(_jar_files()):
+    for _i, jar in enumerate(_jar_files()):
```
Comment (nastra) [inline notebooks/tests/__init__.py:83]: why these changes?

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+   "metadata": {},
+   "source": [
+    "Nessie Spark SQL Demo with NBA Dataset\n",
+    "============================\n",
+    "This demo showcases how to use Nessie Python API along with Spark3 from Iceberg\n",
+    "\n",
+    "Initialize Pyspark\n",
+    "----------------------------------------------\n",
+    "To get started, we will first have to do a few setup steps that give us everything we need\n",
+    "to get started with Nessie. In case you're interested in the detailed setup steps for Spark, you can check out the [docs](https://projectnessie.org/tools/spark/)\n",
```
Comment (nastra) [inline notebooks/nessie-iceberg-demo-nba.ipynb:14]: nit: `.` at the end of the sentence. Same for the the next sentence

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import findspark\n",
+    "from pyspark.sql import *\n",
+    "from pyspark import SparkConf\n",
+    "findspark.init()\n",
```
Comment (nastra) [inline notebooks/nessie-iceberg-demo-nba.ipynb:35]: the demo fails for me on this step with
```
IndexError                                Traceback (most recent call last)
/srv/conda/envs/notebook/lib/python3.7/site-packages/findspark.py in init(spark_home, python_path, edit_rc, edit_profile)
    142     try:
--> 143         py4j = glob(os.path.join(spark_python, "lib", "py4j-*.zip"))[0]
    144     except IndexError:

IndexError: list index out of range

During handling of the above exception, another exception occurred:

Exception                                 Traceback (most recent call last)
<ipython-input-2-9521cc520592> in <module>
      3 from pyspark.sql import *
      4 from pyspark import SparkConf
----> 5 findspark.init()
      6 
      7 conf = SparkConf()

/srv/conda/envs/notebook/lib/python3.7/site-packages/findspark.py in init(spark_home, python_path, edit_rc, edit_profile)
    144     except IndexError:
    145         raise Exception(
--> 146             "Unable to find py4j, your SPARK_HOME may not be configured correctly"
    147         )
    148     sys.path[:0] = [spark_python, py4j]

Exception: Unable to find py4j, your SPARK_HOME may not be configured correctly
```

Am I missing something?

**Review by nastra — CHANGES_REQUESTED** [review]
Demo fails for me using https://hub.gke2.mybinder.org/user/rymurr-nessie-demos-ph3xuepn/notebooks/notebooks/nessie-iceberg-demo-nba.ipynb or maybe I'm missing something?

I also noticed this error when building the Binder image:

```
Installed kernelspec flink-demo in /home/jovyan/.local/share/jupyter/kernels/flink-demo
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/jovyan/notebooks/tests/__init__.py", line 84, in _copy_all_hadoop_jars_to_pyflink
    os.link(jar, pyflink_lib_dir)
FileExistsError: [Errno 17] File exists: '/home/jovyan/hadoop-2.10.1/share/hadoop/common/hadoop-nfs-2.10.1.jar' -> '/srv/conda/envs/flink-demo/lib/python3.7/site-packages/pyflink/lib'
```

File: notebooks/tests/__init__.py (hunk 8 above)
Comment (rymurr) [reply notebooks/tests/__init__.py:83]: flake8 was failing for me locally, not sure why it wasn't picked up by CI

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 9 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:14]: done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 10 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:35]: this is due to the failure w/ the `os.link` that you mentioend above. somehow I didn't catch that in testing. Fixed now.

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+    "conf.set(\"spark.sql.catalog.dev_catalog.uri\", \"http://localhost:19120/api/v1\")\n",
+    "# default branch for Nessie catalog to work on\n",
+    "conf.set(\"spark.sql.catalog.dev_catalog.ref\", \"main\")\n",
+    "# use no authorization. Options are NONE AWS BASIC and aws implies running Nessie on a lambda\n",
+    "conf.set(\"spark.sql.catalog.dev_catalog.auth_type\", \"NONE\")\n",
+    "# enable the extensions for both Nessie and Iceberg\n",
+    "conf.set(\"spark.sql.extensions\", \"org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,org.projectnessie.spark.extensions.NessieSparkSessionExtensions\")\n",
+    "# finally, start up the Spark server\n",
+    "spark = SparkSession.builder.config(conf=conf).getOrCreate()\n",
+    "print(\"Spark Running\")"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:58]: Can you add a `spark.sql("LIST REFERENCES IN dev_catalog").toPandas() ` so the first "real" interaction with Spark/Nessie isn't slow?

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+    },
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "table_count = spark.sql(\"select count(*) from dev_catalog.nba.`salaries@dev`\").toPandas().values[0][0]\n",
+    "csv_count = spark.sql(\"select count(*) from salaries_table\").toPandas().values[0][0]\n",
+    "assert table_count == csv_count\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:180]: Maybe add some `print` of these numbers here, so people have something to look at.

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 13]
```diff
@@ -0,0 +1,629 @@
...
+    "\n",
+    "table_count = spark.sql(\"select count(*) from dev_catalog.nba.`totals_stats@dev`\").toPandas().values[0][0]\n",
+    "csv_count = spark.sql(\"select count(*) from stats_table\").toPandas().values[0][0]\n",
+    "assert table_count == csv_count"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:184]: Same here: Maybe add some `print` of these numbers here, so people have something to look at.

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "spark.sql(\"USE REFERENCE main IN dev_catalog\").toPandas()\n",
+    "try:\n",
+    "    spark.sql(\"SHOW TABLES IN dev_catalog\").toPandas()\n",
+    "except:\n",
+    "    print(\"Error in 0.9.0 of Nessie causes an exception when no tables found\")"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:214]: Can you add the reason for this here?

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 15]
```diff
@@ -0,0 +1,629 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "table_count = spark.sql(\"select count(*) from dev_catalog.nba.salaries\").toPandas().values[0][0]\n",
+    "csv_count = spark.sql(\"select count(*) from salaries_table\").toPandas().values[0][0]\n",
+    "assert table_count == csv_count\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:310]: Same here: Maybe add some print of these numbers here, so people have something to look at.

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 16]
```diff
@@ -0,0 +1,629 @@
...
+    "\n",
+    "table_count = spark.sql(\"select count(*) from dev_catalog.nba.totals_stats\").toPandas().values[0][0]\n",
+    "csv_count = spark.sql(\"select count(*) from stats_table\").toPandas().values[0][0]\n",
+    "assert table_count == csv_count"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:314]: Same here: Maybe add some print of these numbers here, so people have something to look at.

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 17]
```diff
@@ -0,0 +1,629 @@
...
+    "jupyter": {
+     "outputs_hidden": false
+    },
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "spark.sql(\"MERGE BRANCH etl IN dev_catalog\").toPandas()"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:440]: Mind adding an `INTO main` for clarity?

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 18]
```diff
@@ -0,0 +1,629 @@
...
+    "Both branches should be at the same revision after merging/promotion."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "spark.sql(\"MERGE BRANCH dev IN dev_catalog\").toPandas()"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:271]: Mind adding an `INTO main` for clarity?

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 19]
```diff
@@ -0,0 +1,629 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "table_count = spark.sql(\"select count(*) from dev_catalog.nba.allstar_games_stats\").toPandas().values[0][0]\n",
+    "csv_count = spark.sql(\"select count(*) from allstar_table\").toPandas().values[0][0]\n",
+    "assert table_count == csv_count"
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:484]: Same here: Maybe add some `print` of these numbers here, so people have something to look at.

**Review by snazy — COMMENTED** [review]
Mostly LGMT - a couple suggestions for the notebook though

**Review by nastra — DISMISSED** [review]
Overall LGTM, I basically have the same comments as @snazy  re. additional prints

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 12 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:180]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 13 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:184]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 15 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:310]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 16 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:314]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 17 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:440]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 18 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:271]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 19 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:484]: good idea, done

File: notebooks/nessie-iceberg-demo-nba.ipynb (hunk 14 above)
Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:214]: Its a rather complicated interaction b/w the api and the client. I will fix it in an iceberg PR soon but I don't want to burden the user w/ the details.

Comment (rymurr) [reply notebooks/nessie-iceberg-demo-nba.ipynb:214]: For us: its when we set `hashOnRef` and nothing exists on the ref we get a 404

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 20]
```diff
@@ -0,0 +1,634 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "spark.sql(\"USE REFERENCE dev IN dev_catalog\")\n",
+    "\n",
+    "# Creating `salaries` table\n",
+    "spark.sql(\"CREATE TABLE IF NOT EXISTS dev_catalog.nba.salaries (Season STRING, Team STRING, Salary STRING, Player STRING) USING iceberg\")\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:144]:
```suggestion
    "spark.sql(\"\"\"CREATE TABLE IF NOT EXISTS dev_catalog.nba.salaries \n",
    "            (Season STRING, Team STRING, Salary STRING, Player STRING) USING iceberg\"\"\")\n",
```

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 21]
```diff
@@ -0,0 +1,634 @@
...
+    "\n",
+    "spark.sql('CREATE OR REPLACE TEMPORARY VIEW salaries_table USING csv OPTIONS (path \"../datasets/nba/salaries.csv\", header true)')\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:146]:
```suggestion
    "spark.sql(\"\"\"CREATE OR REPLACE TEMPORARY VIEW salaries_table USING csv \n",
    "            OPTIONS (path \"../datasets/nba/salaries.csv\", header true)\"\"\")\n",
```

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 22]
```diff
@@ -0,0 +1,634 @@
...
+    "spark.sql('INSERT INTO dev_catalog.nba.salaries SELECT * FROM salaries_table')\n",
+    "\n",
+    "# Creating `totals_stats` table\n",
+    "spark.sql(\"\"\"CREATE TABLE IF NOT EXISTS dev_catalog.nba.totals_stats (\n",
+    "             Season STRING, Age STRING, Team STRING, ORB STRING, DRB STRING, TRB STRING, AST STRING, STL STRING, \n",
+    "             BLK STRING, TOV STRING, PTS STRING, Player STRING, RSorPO STRING) \n",
+    "             USING iceberg\"\"\")\n",
+    "spark.sql('CREATE OR REPLACE TEMPORARY VIEW stats_table USING csv OPTIONS (path \"../datasets/nba/totals_stats.csv\", header true)')\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:154]:
```suggestion
    "spark.sql(\"\"\"CREATE OR REPLACE TEMPORARY VIEW stats_table USING csv \n",
    "            "OPTIONS (path \"../datasets/nba/totals_stats.csv\", header true)\"\"\")\n",
```

File: notebooks/nessie-iceberg-demo-nba.ipynb [hunk 23]
```diff
@@ -0,0 +1,634 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Creating `allstar_games_stats` table and viewing the contents\n",
+    "spark.sql(\"\"\"CREATE TABLE IF NOT EXISTS dev_catalog.nba.allstar_games_stats (\n",
+    "          Season STRING, Age STRING, Team STRING, ORB STRING, TRB STRING, AST STRING, STL STRING, BLK STRING, \n",
+    "          TOV STRING, PF STRING, PTS STRING, Player STRING) \n",
+    "          USING iceberg\"\"\")\n",
+    "spark.sql('CREATE OR REPLACE TEMPORARY VIEW allstar_table USING csv OPTIONS (path \"../datasets/nba/allstar_games_stats.csv\", header true)')\n",
```
Comment (snazy) [inline notebooks/nessie-iceberg-demo-nba.ipynb:390]:
```suggestion
    "spark.sql(\"\"\"CREATE OR REPLACE TEMPORARY VIEW allstar_table USING csv \n",
    "            OPTIONS (path \"../datasets/nba/allstar_games_stats.csv\", header true)\"\"\")\n",
```

**Comment by snazy** [comment]
Just a few nits (suggestions) regarding horizontal scrolling.

**Review by snazy — APPROVED** [review]
LGTM
